File: superstitioApi/java/superstitioapi/SuperstitioApiSubscriber.java
package superstitioapi;

import basemod.BaseMod;
import basemod.interfaces.*;
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer;
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch2;
import com.evacipated.cardcrawl.modthespire.lib.SpirePrefixPatch;
import com.megacrit.cardcrawl.actions.GameActionManager;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.powers.AbstractPower;
import com.megacrit.cardcrawl.relics.AbstractRelic;
import com.megacrit.cardcrawl.rooms.AbstractRoom;
import superstitioapi.actions.AutoDoneInstantAction;
import superstitioapi.player.PlayerInitPostDungeonInitialize;
import superstitioapi.powers.interfaces.OnPostApplyThisPower;
import superstitioapi.utils.ToolBox;

import java.util.Objects;

import static superstitioapi.InBattleDataManager.ApplyAll;
import static superstitioapi.utils.PowerUtility.foreachPower;


@SpireInitializer
public class SuperstitioApiSubscriber implements
        PostExhaustSubscriber, StartGameSubscriber, RelicGetSubscriber, PostPowerApplySubscriber,
        PostBattleSubscriber, PostDungeonInitializeSubscriber, OnStartBattleSubscriber, OnPlayerTurnStartSubscriber,
        OnCardUseSubscriber, OnPowersModifiedSubscriber, PostDrawSubscriber, PostEnergyRechargeSubscriber, PreMonsterTurnSubscriber {

//    public static boolean hasHadInMonsterTurn = false;

    public SuperstitioApiSubscriber() {
        BaseMod.subscribe(this);
        Logger.run("Done " + this + " subscribing");
    }

    public static void initialize() {
        new SuperstitioApiSubscriber();
    }

    @Override
    public void receivePostExhaust(AbstractCard abstractCard) {
        ApplyAll((sub) -> sub.receivePostExhaust(abstractCard), PostExhaustSubscriber.class);
    }

    @Override
    public void receiveStartGame() {
        ApplyAll(StartGameSubscriber::receiveStartGame, StartGameSubscriber.class);
    }

    @Override
    public void receiveCardUsed(AbstractCard abstractCard) {
        ApplyAll((sub) -> sub.receiveCardUsed(abstractCard), OnCardUseSubscriber.class);
    }

    @Override
    public void receivePowersModified() {
        ApplyAll(OnPowersModifiedSubscriber::receivePowersModified, OnPowersModifiedSubscriber.class);
    }

    @Override
    public void receivePostBattle(AbstractRoom abstractRoom) {
        InBattleDataManager.ClearOnEndOfBattle();
        ApplyAll((sub) -> sub.receivePostBattle(abstractRoom), PostBattleSubscriber.class);
    }

    @Override
    public void receivePostDraw(AbstractCard abstractCard) {
        ApplyAll((sub) -> sub.receivePostDraw(abstractCard), PostDrawSubscriber.class);
    }

    @Override
    public void receivePostDungeonInitialize() {
        ApplyAll(PostDungeonInitializeSubscriber::receivePostDungeonInitialize, PostDungeonInitializeSubscriber.class);
        if (AbstractDungeon.player instanceof PlayerInitPostDungeonInitialize) {
            ((PlayerInitPostDungeonInitialize) AbstractDungeon.player).initPostDungeonInitialize();
        }
    }

    @Override
    public void receivePostEnergyRecharge() {
        ApplyAll(PostEnergyRechargeSubscriber::receivePostEnergyRecharge, PostEnergyRechargeSubscriber.class);
    }

    @Override
    public void receivePostPowerApplySubscriber(
            AbstractPower appliedPower, AbstractCreature target, AbstractCreature source) {
        AutoDoneInstantAction.addToTopAbstract(() -> {
            if (appliedPower instanceof OnPostApplyThisPower)
                target.powers.forEach(power -> {
                    if (power instanceof OnPostApplyThisPower)
                        if (Objects.equals(power.ID, appliedPower.ID)) {
                            ((OnPostApplyThisPower) power).tryInitializePostApplyThisPower(appliedPower, power.getClass());
                        }
                });
        });
        ApplyAll((sub) -> sub.receivePostPowerApplySubscriber(appliedPower, source, source), PostPowerApplySubscriber.class);

    }

    @Override
    public void receiveRelicGet(AbstractRelic abstractRelic) {
        ApplyAll((sub) -> sub.receiveRelicGet(abstractRelic), RelicGetSubscriber.class);
    }

    @Override
    public void receiveOnBattleStart(AbstractRoom abstractRoom) {
        InBattleDataManager.InitializeAtStartOfBattle();
//        hasHadInMonsterTurn = false;
        ApplyAll((sub) -> sub.receiveOnBattleStart(abstractRoom), OnStartBattleSubscriber.class);
    }

    @Override
    public void receiveOnPlayerTurnStart() {
        InBattleDataManager.InitializeAtStartOfTurn();
//        hasHadInMonsterTurn = false;
        ApplyAll(OnPlayerTurnStartSubscriber::receiveOnPlayerTurnStart, OnPlayerTurnStartSubscriber.class);
    }

    @Override
    public boolean receivePreMonsterTurn(AbstractMonster abstractMonster) {
//        if (!hasHadInMonsterTurn)
//            ApplyAll(AtStartOfMonsterTurnSubscriber::atStartOfMonsterTurn, AtStartOfMonsterTurnSubscriber.class);
//        hasHadInMonsterTurn = true;
        ApplyAll((sub) -> sub.receivePreMonsterTurn(abstractMonster), PreMonsterTurnSubscriber.class);
        return true;
    }

//    public interface AtStartOfMonsterTurnSubscriber extends ISubscriber {
//        void atStartOfMonsterTurn();
//    }

    public interface AtEndOfPlayerTurnPreCardSubscriber extends ISubscriber {
        void receiveAtEndOfPlayerTurnPreCard();

        @SpirePatch2(clz = GameActionManager.class, method = "callEndOfTurnActions")
        class AtEndOfTurnSubscriberPatch {
            @SpirePrefixPatch
            public static void Prefix(GameActionManager __instance) {
                ApplyAll(AtEndOfPlayerTurnPreCardSubscriber::receiveAtEndOfPlayerTurnPreCard, AtEndOfPlayerTurnPreCardSubscriber.class);
            }
        }
    }

    public interface AtManualDiscardSubscriber extends ISubscriber {
        void receiveAtManualDiscard();

        interface AtManualDiscardPower {
            void atManualDiscard();
        }

        @SpirePatch2(clz = GameActionManager.class, method = "incrementDiscard", paramtypez = {boolean.class})
        class MotherFuckerWhyTheyDoNotMakeThisDiscardSubscriberPatch {
            @SpirePrefixPatch
            public static void Prefix(boolean endOfTurn) {
                if (endOfTurn) return;
                ApplyAll(AtManualDiscardSubscriber::receiveAtManualDiscard, AtManualDiscardSubscriber.class);
                foreachPower(power ->
                        ToolBox.doIfIsInstance(power, AtManualDiscardPower::atManualDiscard, AtManualDiscardPower.class));
            }
        }
    }


//    public interface AtEndOfRoundSubscriber extends ISubscriber {
//        void receiveAtEndOfRound();
//
//        @SpirePatch2(clz = A.class, method = "applyEndOfTurnTriggers")
//        class AtEndOfRoundSubscriberPatch {
//            @SpirePrefixPatch
//            public static void Prefix(AbstractCreature __instance) {
//                if (__instance instanceof AbstractPlayer)
//                    ApplyAll(AtEndOfRoundSubscriber::receiveAtEndOfRound, AtEndOfRoundSubscriber.class);
//            }
//        }
//    }
}

File: superstitioApi/java/superstitioapi/SuperstitioApiSubscriber.kt
package superstitioapi

import basemod.BaseMod
import basemod.interfaces.*
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch2
import com.evacipated.cardcrawl.modthespire.lib.SpirePrefixPatch
import com.megacrit.cardcrawl.actions.GameActionManager
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import com.megacrit.cardcrawl.powers.AbstractPower
import com.megacrit.cardcrawl.relics.AbstractRelic
import com.megacrit.cardcrawl.rooms.AbstractRoom
import superstitioapi.actions.AutoDoneInstantAction
import superstitioapi.player.PlayerInitPostDungeonInitialize
import superstitioapi.powers.interfaces.OnPostApplyThisPower
import superstitioapi.renderManager.inBattleManager.InBattleDataManager
import superstitioapi.utils.PowerUtility
import superstitioapi.utils.ToolBox

@SpireInitializer
class SuperstitioApiSubscriber : PostExhaustSubscriber, StartGameSubscriber, RelicGetSubscriber,
    PostPowerApplySubscriber, PostBattleSubscriber, PostDungeonInitializeSubscriber, OnStartBattleSubscriber,
    OnPlayerTurnStartSubscriber, OnCardUseSubscriber, OnPowersModifiedSubscriber, PostDrawSubscriber,
    PostEnergyRechargeSubscriber, PreMonsterTurnSubscriber
{
    //    public static boolean hasHadInMonsterTurn = false;
    init
    {
        BaseMod.subscribe(this)
        Logger.run("Done $this subscribing")
    }

    override fun receivePostExhaust(abstractCard: AbstractCard?)
    {
        InBattleDataManager.ApplyAll(
            { sub: PostExhaustSubscriber -> sub.receivePostExhaust(abstractCard) },
            PostExhaustSubscriber::class.java
        )
    }

    override fun receiveStartGame()
    {
        InBattleDataManager.ApplyAll(
            StartGameSubscriber::receiveStartGame,
            StartGameSubscriber::class.java
        )
    }

    override fun receiveCardUsed(abstractCard: AbstractCard?)
    {
        InBattleDataManager.ApplyAll(
            { sub: OnCardUseSubscriber -> sub.receiveCardUsed(abstractCard) },
            OnCardUseSubscriber::class.java
        )
    }

    override fun receivePowersModified()
    {
        InBattleDataManager.ApplyAll(
            OnPowersModifiedSubscriber::receivePowersModified,
            OnPowersModifiedSubscriber::class.java
        )
    }

    override fun receivePostBattle(abstractRoom: AbstractRoom?)
    {
        InBattleDataManager.ClearOnEndOfBattle()
        InBattleDataManager.ApplyAll(
            { sub: PostBattleSubscriber -> sub.receivePostBattle(abstractRoom) },
            PostBattleSubscriber::class.java
        )
    }

    override fun receivePostDraw(abstractCard: AbstractCard?)
    {
        InBattleDataManager.ApplyAll(
            { it.receivePostDraw(abstractCard) },
            PostDrawSubscriber::class.java
        )
    }

    override fun receivePostDungeonInitialize()
    {
        InBattleDataManager.ApplyAll(
            PostDungeonInitializeSubscriber::receivePostDungeonInitialize,
            PostDungeonInitializeSubscriber::class.java
        )
        if (AbstractDungeon.player is PlayerInitPostDungeonInitialize)
        {
            (AbstractDungeon.player as PlayerInitPostDungeonInitialize).initPostDungeonInitialize()
        }
    }

    override fun receivePostEnergyRecharge()
    {
        InBattleDataManager.ApplyAll(
            PostEnergyRechargeSubscriber::receivePostEnergyRecharge,
            PostEnergyRechargeSubscriber::class.java
        )
    }

    override fun receivePostPowerApplySubscriber(
        appliedPower: AbstractPower?, target: AbstractCreature?, source: AbstractCreature?
    )
    {
        if (target == null) return
        AutoDoneInstantAction.addToTopAbstract {
            if (appliedPower is OnPostApplyThisPower<*>)
                for (power: AbstractPower in target.powers)
                {
                    if (power is OnPostApplyThisPower<out AbstractPower>
                        && power.ID == appliedPower.ID
                    )
                        power.tryInitializePostApplyThisPower(appliedPower)
                }
        }
        InBattleDataManager.ApplyAll({ sub: PostPowerApplySubscriber ->
            sub.receivePostPowerApplySubscriber(
                appliedPower,
                source,
                source
            )
        }, PostPowerApplySubscriber::class.java)
    }

    override fun receiveRelicGet(abstractRelic: AbstractRelic?)
    {
        InBattleDataManager.ApplyAll(
            { sub: RelicGetSubscriber -> sub.receiveRelicGet(abstractRelic) },
            RelicGetSubscriber::class.java
        )
    }

    override fun receiveOnBattleStart(abstractRoom: AbstractRoom?)
    {
        InBattleDataManager.InitializeAtStartOfBattle()
        //        hasHadInMonsterTurn = false;
        InBattleDataManager.ApplyAll(
            { sub: OnStartBattleSubscriber -> sub.receiveOnBattleStart(abstractRoom) },
            OnStartBattleSubscriber::class.java
        )
    }

    override fun receiveOnPlayerTurnStart()
    {
        InBattleDataManager.InitializeAtStartOfTurn()
        //        hasHadInMonsterTurn = false;
        InBattleDataManager.ApplyAll(
            OnPlayerTurnStartSubscriber::receiveOnPlayerTurnStart,
            OnPlayerTurnStartSubscriber::class.java
        )
    }

//    override fun receiveOnPlayerDamaged(amount: Int, info: DamageInfo?): Int
//    {
//        var newAmount = amount
//        HangUpCardGroup.forEachHangUpCard(
//            CardOrb_OnAttackedToChangeDamage::class.java
//        ) { orb: CardOrb_OnAttackedToChangeDamage ->
//            newAmount = orb.receiveOnPlayerDamaged(newAmount, info)
//        }
//        return newAmount
//    }

    override fun receivePreMonsterTurn(abstractMonster: AbstractMonster?): Boolean
    {
//        if (!hasHadInMonsterTurn)
//            ApplyAll(AtStartOfMonsterTurnSubscriber::atStartOfMonsterTurn, AtStartOfMonsterTurnSubscriber.class);
//        hasHadInMonsterTurn = true;
        InBattleDataManager.ApplyAll(
            { sub: PreMonsterTurnSubscriber -> sub.receivePreMonsterTurn(abstractMonster) },
            PreMonsterTurnSubscriber::class.java
        )
        return true
    }

    //    public interface AtStartOfMonsterTurnSubscriber extends ISubscriber {
    //        void atStartOfMonsterTurn();
    //    }
    interface AtEndOfPlayerTurnPreCardSubscriber : ISubscriber
    {
        fun receiveAtEndOfPlayerTurnPreCard()

        companion object
        {
            @SpirePatch2(clz = GameActionManager::class, method = "callEndOfTurnActions")
            object AtEndOfTurnSubscriberPatch
            {
                @SpirePrefixPatch
                @JvmStatic
                fun Prefix(__instance: GameActionManager?)
                {
                    InBattleDataManager.ApplyAll(
                        AtEndOfPlayerTurnPreCardSubscriber::receiveAtEndOfPlayerTurnPreCard,
                        AtEndOfPlayerTurnPreCardSubscriber::class.java
                    )
                }
            }
        }
    }

//    interface AtLastHpSubscriber : ISubscriber
//    {
//        fun receiveAtLastHp()
//
//        @SpirePatch(clz = AbstractPlayer::class, method = "damage", paramtypez = [DamageInfo?::class])
//        object OnPlayerDamagedHook
//        {
//            @SpireInsertPatch(localvars = ["damageAmount"], locator = LocatorPre::class)
//            @JvmStatic
//            fun InsertPre(__instance: AbstractPlayer?, info: DamageInfo?, @ByRef damageAmount: IntArray)
//            {
//                var damage = BaseMod.publishOnPlayerDamaged(damageAmount[0], info)
//                if (damage < 0)
//                {
//                    damage = 0
//                }
//
//                damageAmount[0] = damage
//            }
//
//            private class LocatorPre private constructor() : SpireInsertLocator()
//            {
//                @Throws(Exception::class)
//                override fun Locate(ctBehavior: CtBehavior): IntArray
//                {
//                    val matcher: Matcher = MethodCallMatcher(AbstractRelic::class.java, "onLoseHpLast")
//                    return LineFinder.findInOrder(ctBehavior, matcher)
//                }
//            }
//        }
//    }

    interface AtManualDiscardSubscriber : ISubscriber
    {
        fun receiveAtManualDiscard()

        interface AtManualDiscardPower
        {
            fun atManualDiscard()
        }

        companion object
        {
            @SpirePatch2(clz = GameActionManager::class, method = "incrementDiscard", paramtypez = [Boolean::class])
            object MotherFuckerWhyTheyDoNotMakeThisDiscardSubscriberPatch
            {
                @SpirePrefixPatch
                @JvmStatic
                fun Prefix(endOfTurn: Boolean)
                {
                    if (endOfTurn) return
                    InBattleDataManager.ApplyAll(
                        AtManualDiscardSubscriber::receiveAtManualDiscard,
                        AtManualDiscardSubscriber::class.java
                    )
                    PowerUtility.foreachPower { power: AbstractPower ->
                        ToolBox.doIfIsInstance(
                            power,
                            AtManualDiscardPower::class.java,
                            AtManualDiscardPower::atManualDiscard
                        )
                    }
                }
            }
        }
    } //    public interface AtEndOfRoundSubscriber extends ISubscriber {
    //        void receiveAtEndOfRound();
    //
    //        @SpirePatch2(clz = A.class, method = "applyEndOfTurnTriggers")
    //        class AtEndOfRoundSubscriberPatch {
    //            @SpirePrefixPatch
    //            public static void Prefix(AbstractCreature __instance) {
    //                if (__instance instanceof AbstractPlayer)
    //                    ApplyAll(AtEndOfRoundSubscriber::receiveAtEndOfRound, AtEndOfRoundSubscriber.class);
    //            }
    //        }
    //    }

    companion object
    {
        @JvmStatic
        fun initialize()
        {
            val mod = SuperstitioApiSubscriber()
        }
    }


}
